feat: Add Typescript definitions

theogravity opened the pull request:
I've handcrafted TS defs that is based on the API documentation and analysis of the project files.

We have had no problems using it in a project we're working on so far.

I'm unsure if you've worked with Typescript or not, but I've done my best to make sure that for return types where a class is involved, that an interface is returned instead.

This allows for the creation of a mocked class when writing unit tests, vs doing class definition overrides. 

An exception case would be `TokenSet` as it doesn't do anything complicated and is pretty much a plain object under the hood.

panva commented:
Whoa! @theogravity this is great, I will go through the PR to make sure nothing was missed soon.

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+  }
+
+  export class DeviceFlowHandle implements IDeviceFlowHandle {
+    constructor (params : IDeviceFlowHandleParams)
```
remove the constructor from types please, the class is not intended to be constructed manually

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    expired() : boolean
+    expires_at: number
+    client: IClient
+    maxAge: number
```
remove `maxAge`

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    expires_at: number
+    client: IClient
+    maxAge: number
+    exchangeBody: object
```
remove `exchangeBody`

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    client: IClient
+    maxAge: number
+    exchangeBody: object
+    clientAssertionPayload: object
```
remove `clientAssertionPayload`

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    maxAge: number
+    exchangeBody: object
+    clientAssertionPayload: object
+    response: string
```
remove `response`

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    exchangeBody: object
+    clientAssertionPayload: object
+    response: string
+    interval: number
```
remove `interval`

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    verification_uri_complete: string
+    // tslint:disable-next-line:variable-name
+    expires_in: number
+  }
```
same removals as above, in the `DeviceFlowHandle` i don't see the point duplicating everything from the Interface.

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    userinfo_endpoint: string
+    revocation_endpoint: string
+    introspection_endpoint: string
+  }
```
add `device_authorization_endpoint`

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    mtls_endpoint_aliases: IMetadataMtlsEndpointAliases
+
+    [key: string]: any
+  }
```
add `device_authorization_endpoint`

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    /**
+     * State value passed in the authentication request
+     */
+    session_state?: string | object
```
never an object.

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    // tslint:disable-next-line:variable-name
+    expires_at?: number
+    // tslint:disable-next-line:variable-name
+    session_state?: string | object
```
never an object

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    introspection_endpoint_auth_signing_alg_values_supported: string
+    revocation_endpoint_auth_methods_supported: string
+    revocation_endpoint_auth_signing_alg_values_supported: string
+    request_object_signing_alg_values_supported: string
```
any property ending with `_supported` is a `string[]`, not a string.

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    /**
+     * Returns the <Client> class tied to this issuer.
+     */
+    Client: { new (metadata: IClientMetadata, jwks?: object) }
```
dtto as `jwks` above.

panva commented:
@ulrichb @YangusKhan @RMHonor as the three contributors to the DefinitelyTyped types, i would appreciate your review here as well.

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+  /**
+   * @see https://medium.com/@darutk/diagrams-of-all-the-openid-connect-flows-6968e3990660
+   */
+  enum AuthResponseType {
```
If we're defining ambient enums in the declaration file, they'll also need to exist in the javascript to be useable

RMHonor replied:
May be more suitable to use a string literal if we want to restrict the set of auth types without needing to access an existing enum in JS

panva replied:
Agreed, ergo please change all enums to `type Foo = 'a' | 'b' | 'c'` instead

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+  /**
+   * @see https://github.com/panva/node-openid-client/blob/master/docs/README.md#client-authentication-methods
+   */
+  enum TokenAuthMethod {
```
If we're defining ambient enums in the declaration file, they'll also need to exist in the javascript to be useable

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    revocation_endpoint_auth_signing_alg?: string
+    tls_client_certificate_bound_access_tokens?: boolean
+
+    [key: string]: any
```
`unknown` is a safer type to use here, so the developer has to explicitly cast it to what they believe the type should be

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    redirect_uri?: string
+    response_type?: string
+    scope?: string,
+    [key: string]: any
```
`unknown` is a safer type to use here, so the developer has to explicitly cast it to what they believe the type should be

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    response_type?: string
+    nonce?: string
+    code_verifier?: string
+    [key: string]: any
```
`unknown` is a safer type to use here, so the developer has to explicitly cast it to what they believe the type should be

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+
+  export interface IGrantBody {
+    grant_type: string
+    [key: string]: any
```
`unknown` is a safer type to use here, so the developer has to explicitly cast it to what they believe the type should be

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    iat?: number
+    exp?: number
+    jti?: string
+    [key: string]: any
```
`unknown` is a safer type to use here, so the developer has to explicitly cast it to what they believe the type should be

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+  export interface IDeviceAuthParameters {
+    client_id?: string
+    scope?: string
+    [key: string]: any
```
`unknown` is a safer type to use here, so the developer has to explicitly cast it to what they believe the type should be

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+     */
+    deviceAuthorization(parameters?: IDeviceAuthParameters, extras?: IDeviceAuthExtras): Promise<IDeviceFlowHandle>
+
+    [key: string]: any
```
Are you sure we want this extension here? We shouldn't be assuming there are additional properties on a client

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+    request_object_signing_alg_values_supported: string
+    mtls_endpoint_aliases: IMetadataMtlsEndpointAliases
+
+    [key: string]: any
```
Ditto with `unknown`

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+     */
+    keystore (forceReload?: boolean): Promise<JWKS.KeyStore>
+
+    [key: string]: any
```
As above, should an `Issuer` allow additional types like this?

theogravity replied:
My linter fails if it's not included since it's implemented in `IIssuer`

RMHonor commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+     */
+    session_state?: string | object
+
+    [key: string]: any
```
`unknown` again

RMHonor replied:
on second thought, should we allow this to be extensible at all? I can't think of an instance where a `TokenSet` will have additional properties, @panva may be better placed to comment

panva replied:
Anything returned by the token response is available actually. So yeah. But as you suggest above, unknown as a type.

big-kahuna-burger commented:
Can we also have `Strategy` exported for the sake of completeness?

panva commented:
I'd like someone to write .ts test file using the types so that we may have a CI job verifying them. Any takers?

big-kahuna-burger commented:
> I'd like someone to write .ts test file using the types so that we may have a CI job verifying them. Any takers?

@panva I will take it (via `tsd`). Expect PR later in the evening.

Bessonov commented:
Great! While it's better to have definitions instead of not to have definitions at all... wouldn't it better to convert the project to typescript? What do you think @panva ? Maybe step by step.

> I'd like someone to write .ts test file using the types so that we may have a CI job verifying them.

I'm not sure that this would help to check definitions. AFAIK you will check fictional*1 definitions with fictional*2 tests.

*1 fictional, because they can be not related to concrete implementation.
*2 fictional, because it doesn't make much sense to create high coverage tests just for checking typings.

Maybe it's better to convert tests to ts as a first step?

panva commented:
Let's park the tests for now until we're satisfied with the state of the definition.

@Bessonov exposing types is as far as I'm willing to go at this point in time.

theogravity commented on lib/index.d.ts:
```diff
@@ -0,0 +1,528 @@
+  }
+
+  export class Client implements IClient {
+    constructor (metadata: IClientMetadata, jwks?: object)
```
sure

theogravity commented:
Thank you everyone for the quick turnaround!

I've addressed the most recent comments - apply `JSONWebKeySet`, and remove the `DeviceFlowHandle` def.

panva commented on lib/index.d.ts:
```diff
@@ -186,11 +186,11 @@ declare module 'openid-client' {
      * Initial Access Token to use as a Bearer token during the registration call.
      */
-    initialAccessToken?: string
+    initialAccessToken: string
```
this is still optional

panva commented on lib/index.d.ts:
```diff
@@ -186,11 +186,11 @@ declare module 'openid-client' {
      * When neither jwks_uri or jwks is present in metadata the key's public parts will be registered as jwks.
      */
-    jwks?: object
+    jwks: JSONWebKeySet
```
this is still optional

panva commented:
@theogravity thank!

Re-iterating on the rest of missing pieces

> In addition to those we neeed to add [generators](https://github.com/panva/node-openid-client/tree/master/docs#generators) and the [customization](https://github.com/panva/node-openid-client/tree/master/docs#customizing) functionality, some of this is already part of the community maintained [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/openid-client/index.d.ts).

I appreciate the effort you've put into this.

theogravity commented:
Todo:
-`Strategy`
-`Errors`
- `generators`
- customization

theogravity commented on lib/index.d.ts:
```diff
@@ -485,4 +487,29 @@ declare module 'openid-client' {
+    usePKCE: boolean | string
+  }
+
+  export type StrategyVerifyCallbackUserInfo<TUser> = (tokenset: TokenSet, userinfo?: object, done?: (err: any, user?: TUser) => void) => void)
```
I had to make `done` optional because typescript will not allow an optional parameter followed by a required one.

Bessonov replied:
If I read [it correctly](https://github.com/panva/node-openid-client/blob/master/docs/README.md#new-strategyoptions-verify) (I'm not familiar with it), then neither `done` nor `userinfo` should be optional in this type. For use case without `userinfo` you have the type `StrategyVerifyCallback` below. Or do I'm missing something?

theogravity replied:
`done` should not be optional, but because `userinfo` is, doing the following is illegal in TS:

`fn (userinfo?, done)`

I looked at the tests for the `Strategy` class in this project, and saw two usages for the verify callback:

`(tokenset, done)
`(tokenset, userinfo, done)`

The docs has `userinfo` as optional, and the tests for the `Strategy` also reflects both usages as well.

Bessonov replied:
But for this reason you have:
```ts
verify: StrategyVerifyCallbackUserInfo<TUser> | StrategyVerifyCallback<TUser>
```

Then it should be enough:
```ts
export type StrategyVerifyCallbackUserInfo<TUser> = (tokenset: TokenSet, userinfo: object, done: (err: any, user?: TUser) => void) => void
export type StrategyVerifyCallback<TUser> = (tokenset: TokenSet, done: (err: any, user?: TUser) => void) => void
```

theogravity replied:
Oh, I see. Thanks! Will update.

theogravity commented:
That should complete the definitions!

I based the error defs on what I saw here:
https://github.com/panva/node-openid-client/blob/master/lib/errors.js

Feel free to update it if it looks incorrect in any way

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,644 @@
+   * @see https://github.com/panva/node-openid-client/blob/master/lib/index.js
+   */
+  export const custom : {
+    setHttpOptionsDefaults(params: ISetHttpOptionsDefaults) : void
```
params here is actually following the `GotOptions` interface too

theogravity replied:
ok

panva commented on lib/index.d.ts:
```diff
@@ -0,0 +1,644 @@
+/// <reference types="@panva/jose" />
```
someone care to explain to a newbie here why this needs to be referenced but `got` doesn't?

theogravity replied:
It should be included. I'll update.

https://stackoverflow.com/questions/48216163/what-does-reference-types-node-mean

theogravity commented:
The CI needs a kick on one of them - been stalled for hours on npm install

panva commented:
I’ll retrigger with an empty push, UI doesn’t seem to work all that well ;)

panva commented:
One more thing - the inline docs in the ts definition - does it show up in intellisence for developers? In my experience it doesn't so i'd rather remove it and only have regular docs to avoid having to maintain two places?

theogravity commented:
Just realized - I was testing in my own project, but since this is to be included in yours, the module declaration isn't necessary.

I was able to get comments in VS Code after doing that. Try again.

If you don't get the `got` options, try `npm i @types/got --save-dev`

panva commented:
@ulrichb @YangusKhan @RMHonor @big-kahuna-burger @Bessonov 

I'm looking for more ✅ from you guys, please `npm i theogravity/node-openid-client#typescript-defs` and check if this fits to replace @types/openid-client

theogravity commented on lib/index.d.ts:
```diff
@@ -638,5 +640,9 @@ export namespace errors {
   export class RPError extends Error {
     response?: IncomingMessage
+    jwt?: string
+    jwt?: string
```
you have jwt defined twice

big-kahuna-burger commented:
> @ulrichb @YangusKhan @RMHonor @big-kahuna-burger @Bessonov
> 
> I'm looking for more ✅ from you guys, please `npm i theogravity/node-openid-client#typescript-defs` and check if this fits to replace @types/openid-client

LGTM, Strategy is in which I wished for. Thanks @theogravity 🚀

ulrichb commented on lib/index.d.ts:
```diff
@@ -0,0 +1,648 @@
+  /**
+   * Introspects a token at the Authorization Server's introspection_endpoint.
+   */
+  introspect(token: string, tokenTypeHint?: string, extras?: IIntrospectExtras): Promise<object>
```
Here a more detailed response type would bee cool.

For example `active: boolean` is a required response property. See  https://tools.ietf.org/html/rfc7662#section-2.2.

Maybe just copy [`IntrospectionResponse`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d8907f303d75056e0710a72bad713179fbc78240/types/openid-client/index.d.ts#L93)?

panva replied:
better now?

ulrichb commented on lib/index.d.ts:
```diff
@@ -0,0 +1,648 @@
+   * @param accessToken Access Token value. When TokenSet instance is provided its access_token property
+   * will be used automatically.
+   */
+  userinfo(accessToken: TokenType): Promise<object>
```
Please use `{ readonly [name: string]: unknown }` as return type instad of `object`.

This is still type safe (because of the `unknown` type) and avoids reference errors.

See [this sample](https://www.typescriptlang.org/play/#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzg2RlQHkAjAK3AwAoBKeALnhypoG4BYAKDDwDOOCCAB0EHAHNahYmXaYGogcnL1uPXqEiwEKdNjwEQREhWqYA6lgwALAJKpQADxAwGzeAG9jUYHggAT3gAbVQoAFsQFgEMGCxUSQBdFjQAa1QcAHd8AF8NflQhEXEpGRM5cxprO0cXNyUVNQ0gA) for the difference:

![image](https://user-images.githubusercontent.com/388796/64430911-9511a700-d0b9-11e9-8f86-194c7938df7f.png)

panva replied:
better now?

ulrichb commented:
... also unit test (+ execution in the build script) would be cool.

As a starting point you could use these: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/openid-client/openid-client-tests.ts

panva commented:
> ... also unit test (+ execution in the build script) would be cool.
> 
> As a starting point you could use these: [DefinitelyTyped/DefinitelyTyped:types/openid-client/openid-client-tests.ts@`master`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/openid-client/openid-client-tests.ts) [](/DefinitelyTyped/DefinitelyTyped/blob/HEAD@{2019-09-06T13:25:23Z}/types/openid-client/openid-client-tests.ts)

@ulrichb can you elaborate on the steps necessary to execute this?

theogravity commented:
Based on reading the DefinitelyTyped repo:

You write a series of statements, such as:

```
const client = new Client(...)
```

And you write enough of them that it would cover the types that you've written in your definition, then you execute a TS linter against it. If the linter passes, then it's all good.

So what I'm guessing here is if you want to actually have the CI test the defs file, is you write a sample `.ts` file and run the TS linter on it.

theogravity commented:
Let me know if this is something you'd like me to take the lead on, I'd be happy to set it up, just didn't want to mess with your dependencies if possible.

panva commented:
I refactored the types, removed a lot of the clutter and added the test to CI (part of the `Lint` job).

Please take a look - @big-kahuna-burger i made changes to the strategy definition, can you please verify it still works as intended for you? I didn't bother with the tests yet.

theogravity commented:
I would really recommend that a class extends an interface and all methods that take in or return a class, use the interface instead 

from a unit test perspective, if it's a class, then i generally need to be able to create an instance of that class and create spies against it or override its prototype to do any custom implementation

vs if it's an interface, i can generally create a custom class that implements the interface or extends a class that implements it and overuse the methods directly

theogravity commented:
I would really recommend that anywhere a class is defined as part of a property, parameter, return, etc, that a generic be used. It would make life so much easier when unit testing as I'd be able to create my own custom class and easily override any methods that I need to mock.

Not only that, if I end up writing a custom class that extends it and add new methods, typescript would be able to provide autocompletion support, refactor capabilities, and would recognize my new method because I can specify my custom class as the type for the generic

```
class Issuer <C extends Client> {
     client: C

     constructor(client : C)
}

function blah<C extends Client, I extends Issuer>(client: T, issuer: I) {

}

class CustomIssuer extends Issuer {
   myCustomFn() {

   }
}

function blah<Client, CustomIssuer>(new Client(), new CustomIssuer()) {
   // this wouldn't be valid without generics
   issuer.myCustomFn()
}
```

panva commented:
@theogravity as-is now i can not only understand the types but first and foremost can continue *maintaining* them. Not using TS myself I don't get the samples you're showing me.

Furthermore the types that the community published to DefinitelyTyped are closer to what the state is now and it seems to be enough.

What do others think?

theogravity commented:
I'm unsure if your comment was directed towards my comment about introducing back the interface, but I deleted the comment because I realized that maintainability / duplication would be an issue.

a compromise that still works well is using generics instead - what are your thoughts on generics?

panva commented:
It is directed at your [comment](https://github.com/panva/node-openid-client/pull/184#issuecomment-529130541). "generics" has no meaning to me as non-TS user. If you want to show me please make the update. But no promises, it'll stick :)

theogravity commented:
Understood. Will do it right now.

panva commented:
> Understood. Will do it right now.

no rush, i ain't got time today no more.

theogravity commented:
No problem. Thanks for refactoring it. 

I've added generics where it made sense to. Hopefully the tests demonstrates how powerful it can be. 

I had to add the types package for passport-strategy and got as VS Code was not recognizing the types at all until I added them.

panva commented:
Thank you all for feedback and contributions. This has landed in 3.7.0

ulrichb commented:
If the consuming project doesn't have `@types/got` installed, you get:

> error TS7016: Could not find a declaration file for module 'got'.

=> Should we move `@types/got` from the dev-dependencies to the dependencies, so that TS compilation works out of the box?

panva commented:
I intentionally did not do that, I don't see why non-ts users should have this as a dependency.

Maybe add a typescript readme section with this detail? Is that a big deal? Aren't ts users used to installing types? I assumed you do.

RMHonor commented:
My first instinct was to put it in peer dependencies, as it then at least notifies the user. But perhaps optional dependencies would be more appropriate? Although it's not an approach I've encountered before

panva commented:
@RMHonor neither of these two applies

optionalDependencies

> If a dependency can be used, but you would like npm to proceed if it cannot be found or fails to install, then you may put it in the optionalDependencies object.

peerDependencies

> In some cases, you want to express the compatibility of your package with a host tool or library, while not necessarily doing a require of this host.

ulrichb commented:
@panva The (small) problem exists if you want to use `node-openid-client`, but don't use `got` in the consuming project. Then a) you first get the compile error mentioned above and b) you have to add an unnecessary `@types\got` package dependency in the consuming project.

[Optional dependency](https://npm.github.io/using-pkgs-docs/package-json/types/optionaldependencies.html) looks like a good option :).

panva commented:
> Optional dependency looks like a good option :).

It isn't, it still installs for everyone. It just that it can optionally fail to install.

panva commented:
> you have to add an unnecessary @types\got package dependency in the consuming project.

How is it unnecessary? If you're using TS you need those types. If you don't you don't have to install it.

RMHonor commented:
Something like `optionalPeerDependencies` would be perfect. I'd be inclined to say that it's fine to leave if as it is, and a small brief in the documentation around installing the 3rd party type definitions.

ulrichb commented:
@panva: 
> How is it unnecessary?

If you don't use `got` directly in the consuming project (only indirectly via `node-openid-client`). Still without having `@types\got` installed you immediately get a compile error in this line:

https://github.com/panva/node-openid-client/blob/45402ede49375ed9bd9660f39d544f4483340bff/types/index.d.ts#L10

=> "error TS7016: Could not find a declaration file for module 'got'"

panva commented:
In that case, i'm keen to remove the type dependency altogether and make it 

```
{
  [key: string]: unknown;
}
```

and jsdoc add `@see https://github.com/sindresorhus/got/tree/v9.6.0#options`

ulrichb commented:
Why is a normal dependency so bad? Packages often pull dependencies which aren't used in all cases / for all users.

An example: `express` has a dependency to `cookie` (or `cookie-signature`, or `serve-static`) which obviously isn't used by every `express` user.

ulrichb commented:
Another argument: Also other packages to it that way.

For example `react-intl` (just an example; it was the first one I found in my project):

https://github.com/formatjs/react-intl/blob/899d17239545717283b1d539b011240428e9948d/package.json#L39-L40

panva commented:
> Why is a normal dependency so bad? Packages often pull dependencies which aren't used in all cases / for all users.

Packages often do, I choose not to. @types/got has three dependencies which also pack other transitive dependencies.

Showcasing bloated packages that do this is not an argument if i'm honest.

I was about to add the dependency then i checked what it means in terms of other transitive dependencies, so instead if someone could contribute the removal of that import altogether replacing it with the actual object with properties that make sense.

panva commented:
v3.7.1 fixes this.

ulrichb commented:
>  I'll move forward with the removal of that import altogether replacing it with the actual object with properties that make sense.

👍

ulrichb commented:
Just tested it. Compiles fine without `@types\got` installed.

Many thanks @panva!

theogravity commented:
Thank you everyone!

RMHonor commented:
Thanks a lot @theogravity, really appreciate these definitions
